File: 377. Combination Sum IV.swift
import Darwin

// RunTime Error

//func combinationSum4(nums: [Int], _ target: Int) -> Int {
//    let nums = nums.sort()
//    var res = Array(count: target + 1, repeatedValue: 0)
//    res[0] = 1
//    for i in 1...target {
//        for n in nums {
//            if n <= i {
//                res[i] += res[i - n]
//            }else {
//                break
//            }
//        }
//    }
//    return res[target]
//}

let nums = [3, 33, 333], target = 10000

combinationSum4(nums, target)

File: 377. Combination Sum IV.cpp
// 4 ms

class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        std::sort(nums.begin(), nums.end());
        vector<int> res(target + 1);
        res[0] = 1;
        for (int i = 1; i <= target; ++i) {
            for (int n : nums) {
                if (n > i) {
                    break;
                }
                if (n <= i) {
                    res[i] += res[i-n];
                }
            }
        }
        return res[target];
    }
};